core/envelope-sim/src/main/java/fr/sncf/osrd/envelope_sim/EnvelopeSimPath.java

```diff
@@ -135,25 +136,58 @@ public void setElectricalProfiles(Map<String, RangeMap<Double, String>> electric
+            TreeRangeMap<Double, ElecCond> profilesAndModes = TreeRangeMap.create();
             profilesAndModes.putAll(catenaryModeMap);
 
             for (var profileEntry : profileMap.asMapOfRanges().entrySet()) {
                 for (var entry : catenaryModeMap.subRangeMap(profileEntry.getKey()).asMapOfRanges().entrySet()) {
                     var mode = entry.getValue().mode();
                     var profileLevel = profileEntry.getValue();
-                    profilesAndModes.put(entry.getKey(), new ModeAndProfile(mode, profileLevel));
+                    profilesAndModes.put(entry.getKey(), new ElecCond(mode, profileLevel, null));
                 }
             }
             modeAndProfileMapsByPowerClass.put(powerClassEntry.getKey(), ImmutableRangeMap.copyOf(profilesAndModes));
         }
     }
 
-    public RangeMap<Double, ModeAndProfile> getModeAndProfileMap(String powerClass) {
-        if (modeAndProfileMapsByPowerClass == null)
-            return catenaryModeMap;
-        return modeAndProfileMapsByPowerClass.getOrDefault(powerClass, catenaryModeMap);
+    private RangeMap<Double, ElecCond> getModeAndProfileMap(String powerClass, Range<Double> range,
+                                                           boolean ignoreElectricalProfiles) {
+        if (modeAndProfileMapsByPowerClass == null || ignoreElectricalProfiles)
+            return catenaryModeMap.subRangeMap(range);
+        return modeAndProfileMapsByPowerClass.getOrDefault(powerClass, catenaryModeMap).subRangeMap(range);
+    }
+
+    /**
+     * Get the catenary related data for a given power class and power restriction map.
+     */
+    public ImmutableRangeMap<Double, ElecCond> getElecCondMap(String basePowerClass,
+                                                     RangeMap<Double, String> powerRestrictionMap,
+                                                     Map<String, String> powerRestrictionToPowerClass,
+                                                     boolean ignoreElectricalProfiles) {
+        TreeRangeMap<Double, ElecCond> res = TreeRangeMap.create();
+        res.putAll(getModeAndProfileMap(basePowerClass, Range.closed(0.0, length), ignoreElectricalProfiles));
+
+        if (powerRestrictionMap != null) {
+            for (var entry : powerRestrictionMap.asMapOfRanges().entrySet()) {
+                var powerClass = powerRestrictionToPowerClass.getOrDefault(entry.getValue(), basePowerClass);
+                var modeAndProfileMap = getModeAndProfileMap(powerClass, entry.getKey(), ignoreElectricalProfiles);
+                for (var modeAndProfileEntry : modeAndProfileMap.asMapOfRanges().entrySet()) {
+                    var mode = modeAndProfileEntry.getValue().mode();
+                    var profile = modeAndProfileEntry.getValue().profile();
+                    res.putCoalescing(modeAndProfileEntry.getKey(), new ElecCond(mode, profile, entry.getValue()));
+                }
+            }
```

Comment:
We "merge" the power restriction map with the mode and profile map any time the electrical condition map is requested? That sounds expensive.

Is there a reason why the power restriction data isn't part of the path?